### PR TITLE
refactor: worker errors

### PR DESCRIPTION
**Pull request by mehulmathur16**

### CODE REVIEW

File: src/cli/javascript/runtime.rs [hunk 1]
```diff
@@ -82,37 +83,49 @@ impl Drop for Runtime {
...
 }
 
 #[async_trait::async_trait]
 impl WorkerIO<ConstValue, ConstValue> for Runtime {
-    async fn call(&self, name: &str, input: ConstValue) -> anyhow::Result<Option<ConstValue>> {
+    async fn call(
+        &self,
+        name: &str,
+        input: ConstValue,
+    ) -> crate::core::Result<Option<ConstValue>, worker::Error> {
```
Comment (tusharmath) [inline src/cli/javascript/runtime.rs:113]:
```suggestion
    ) -> worker::Result<Option<ConstValue>> {
```

File: src/cli/javascript/runtime.rs [hunk 2]
```diff
@@ -82,37 +83,49 @@ impl Drop for Runtime {
...
         let script = self.script.clone();
         let name = name.to_string();
-        let value = serde_json::to_string(&input)?;
+        let value = serde_json::to_string(&input).map_err(worker::Error::from)?;
```
Comment (tusharmath) [inline src/cli/javascript/runtime.rs:116]:
```suggestion
        let value = serde_json::to_string(&input)?;
```
The question mark should automatically convert.

File: src/cli/javascript/runtime.rs [hunk 3]
```diff
@@ -82,37 +83,49 @@ impl Drop for Runtime {
...
         if let Some(runtime) = &self.tokio_runtime {
             runtime
                 .spawn(async move {
                     init_rt(script)?;
-                    execute_inner(name, value).map(Some)
+                    execute_inner(name, value)
+                        .map(Some)
+                        .map_err(worker::Error::from)
```
Comment (tusharmath) [inline src/cli/javascript/runtime.rs:123]:
```suggestion
                        .map(Some)?
```

File: src/cli/javascript/runtime.rs [hunk 4]
```diff
@@ -82,37 +83,49 @@ impl Drop for Runtime {
...
                 })
-                .await?
+                .await
+                .map_err(|_| worker::Error::ExecutionFailed)?
```
Comment (tusharmath) [inline src/cli/javascript/runtime.rs:126]:
```suggestion
                .await?
```

File: src/cli/javascript/codec.rs [hunk 5]
```diff
@@ -214,18 +214,19 @@ mod test {
...
+    fn create_test_response() -> Result<WorkerResponse, worker::Error> {
         let mut headers = HeaderMap::new();
         headers.insert("content-type", "application/json".parse().unwrap());
         let response = crate::core::http::Response {
             status: reqwest::StatusCode::OK,
             headers,
             body: Bytes::from("Hello, World!"),
         };
-        let js_response: Result<WorkerResponse> = response.try_into();
+        let js_response: Result<WorkerResponse, worker::Error> = response.try_into();
```
Comment (tusharmath) [inline src/cli/javascript/codec.rs:229]:
```suggestion
        let js_response: worker::Result<WorkerResponse> = response.try_into();
```

File: src/core/ir/error.rs [hunk 6]
```diff
@@ -28,6 +28,9 @@ pub enum Error {
 
     #[error("Authentication Failure: {0}")]
     AuthError(auth::error::Error),
+
+    #[error("Worker Error: {0}")]
+    WorkerError(String),
```
Comment (tusharmath) [inline src/core/ir/error.rs:33]:
```suggestion
    WorkerError(worker::Error),
```

Comment (tusharmath) [reply src/core/ir/error.rs:33]: We can use `derive_more` to implement `From`

File: src/core/ir/error.rs [hunk 7]
```diff
@@ -87,3 +90,9 @@ impl From<anyhow::Error> for Error {
         }
     }
 }
+
+impl From<error::worker::Error> for Error {
+    fn from(value: error::worker::Error) -> Self {
+        Error::WorkerError(value.to_string())
+    }
+}
```
Comment (tusharmath) [inline src/core/ir/error.rs:98]: Not required.

File: src/core/mod.rs [hunk 8]
```diff
@@ -89,7 +91,7 @@ pub type EntityCache = dyn Cache<Key = IoId, Value = ConstValue>;
 #[async_trait::async_trait]
 pub trait WorkerIO<In, Out>: Send + Sync + 'static {
     /// Calls a global JS function
-    async fn call(&self, name: &str, input: In) -> anyhow::Result<Option<Out>>;
+    async fn call(&self, name: &str, input: In) -> Result<Option<Out>, error::worker::Error>;
```
Comment (tusharmath) [inline src/core/mod.rs:94]:
```suggestion
    async fn call(&self, name: &str, input: In) -> worker::Result<Option<Out>>;
```

File: src/core/error.rs [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+
+use derive_more::From;
+
+#[derive(From, thiserror::Error, Debug)]
+pub enum Error {
+    #[error("Worker Error")]
+    Worker(worker::Error),
+}
+
+pub mod worker {
```
Comment (tusharmath) [inline src/core/error.rs:11]: Move this to `worker::error` in `worker.rs`

File: src/cli/javascript/runtime.rs [hunk 10]
```diff
@@ -139,50 +140,48 @@ fn prepare_args<'js>(ctx: &Ctx<'js>, req: WorkerRequest) -> rquickjs::Result<(Va
...
-            .get_mut()
-            .ok_or(anyhow::anyhow!("JS runtime not initialized"))?;
+        let runtime = cell.get_mut().ok_or(worker::Error::RuntimeNotInitialized)?;
         runtime.0.with(|ctx| match event {
             Event::Request(req) => {
                 let fn_as_value = ctx
                     .globals()
                     .get::<&str, Function>(name.as_str())
-                    .map_err(|_| anyhow::anyhow!("globalThis not initialized"))?;
+                    .map_err(|_| worker::Error::GlobalThisNotInitialised)?;
```
Comment (tusharmath) [inline src/cli/javascript/runtime.rs:151]:
```suggestion
                    .map_err(|e| worker::Error::GlobalThisNotInitialised(e))?;
```

Comment (tusharmath) [reply src/cli/javascript/runtime.rs:151]: If we implement `from` then we might not even need `map_err`

File: src/cli/javascript/runtime.rs [hunk 11]
```diff
@@ -139,50 +140,48 @@ fn prepare_args<'js>(ctx: &Ctx<'js>, req: WorkerRequest) -> rquickjs::Result<(Va
...
 
-                let args = prepare_args(&ctx, req)?;
+                let args =
+                    prepare_args(&ctx, req).map_err(|e| worker::Error::Rquickjs(e.to_string()))?;
                 let command: Option<Value> = function.call(args).ok();
                 command
                     .map(|output| Command::from_js(&ctx, output))
                     .transpose()
-                    .map_err(|e| anyhow::anyhow!("deserialize failed: {e}"))
+                    .map_err(|e| worker::Error::DeserializeFailed(e.to_string()))
```
Comment (tusharmath) [inline src/cli/javascript/runtime.rs:163]: Shoud use `?` and leveraged the derived From.

File: src/cli/javascript/runtime.rs [hunk 12]
```diff
@@ -139,50 +140,48 @@ fn prepare_args<'js>(ctx: &Ctx<'js>, req: WorkerRequest) -> rquickjs::Result<(Va
...
             let function = fn_as_value
                 .as_function()
-                .ok_or(anyhow::anyhow!("`{}` is not a function", name))?;
-            let val: String = function.call((value, ))
-                .map_err(|_| anyhow::anyhow!("unable to parse value from js function: {} maybe because it's not returning a string?", name,))?;
-            Ok::<_, anyhow::Error>(serde_json::from_str(&val)?)
+                .ok_or(worker::Error::InvalidFunction(name.clone()))?;
+            let val: String = function
+                .call((value,))
+                .map_err(|_| worker::Error::FunctionValueParseError(name))?;
```
Comment (tusharmath) [inline src/cli/javascript/runtime.rs:183]: Loosing information about the error here.

**Comment by mehulmathur16** [comment]
@tusharmath I have used map_err for handling rquickjs::Error, Since it can't be imported in core/. And worker::Error exists in core/error.rs

**Review by tusharmath — APPROVED** [review]

**Pull request merged**